lib/spack/spack/solver/concretize.lp

```diff
@@ -1077,6 +1088,54 @@ variant_is_propagated(PackageNode, Variant) :-
    attr("variant_selected", PackageNode, Variant, Value, _, _),
    not propagate(PackageNode, variant_value(Variant, Value)).
 
+% a node is the source of propagation if nothing propagated to it
+propagation_source(Source, Variant, Value) :-
+  propagate(Source, variant_value(Variant, Value)),
+  not 1 { propagate(Parent, variant_value(Variant, Value)): depends_on(Parent, Source) }.
```

Comment:
(Not necessarily for this PR) I think this can be improved, in the sense that we know from user input which are the sources of propagation, and we don't need to have cardinality conditions taking into account all possible deps to recompute that here.

Reply:
https://github.com/spack/spack/pull/42931/commits/61b3ba3ceefbd914caf76af6d484816f313dcd0f